File: typescript/csv_parser.ts

import * as csv from 'csv-parser';
import * as fs from 'fs';
import * as json from 'jsonfile';
import * as pg from 'pg';
import * as request from 'request';

interface ApiRow {
    id: number;
    url: string;
    name: string;
    created: number;
}

function printTableApi(dbConnString: string): void {

        const client = new pg.Client(dbConnString);

        client.connect();

        client.query('SELECT * FROM api;', (err, res) => {
            if (err) throw err;

            const rows: ApiRow[] = res.rows;    
            const columns = res.fields.map(field => field.name);    

            console.log(`Table: api`);    
            rows.forEach(row => {    
                columns.forEach((column, index) => {   
                console.log(`${column}: ${row[column]} `);     
                });    
            console.log('');    
        });    

        client.end(); 

        });
}

function processCsvData(url: string, dbConnString: string): void {

        request.get(url)
        .on('error', err => { throw err; })
        .pipe(csv())
        .on('data', (row: any) => {
            const client = new pg.Client(dbConnString);
            client.connect();

        client.query(`    
            CREATE TABLE IF NOT EXISTS api (    
            id SERIAL PRIMARY KEY,    
            url TEXT,    
            name TEXT,    
            created INTEGER    
         );    
        `,
        (err, res) => {    
                 if (err) throw err;    

            client.query(`    
                INSERT INTO api (url, name, created)    
                 VALUES ($1, $2, $3);    
             `, [row.url, row.name, row.created], (err, res) => {    
                 if (err) throw err;    

                    client.end();    
             });    
        });    
    })    
    .on('end', () => {    
        console.log('Data inserted successfully!');    
    });  
      
}

async function apiHandler(): Promise<string> {
    const dbConnString = 'postgres://postgres:postgres@localhost:5432/data';
    const client = new pg.Client(dbConnString);
    await client.connect();

    try {
     const res = await client.query('SELECT * FROM api;')
     const rows: ApiRow[] = res.rows;    
     const columns = res.fields.map(field => field.name);    

    const apis = rows.map(row => {    
          const api = {};    
          columns.forEach((column, index) => {    
          api[column] = row[column];    
     });  

      return api;  

    });    
        await  client.end(); 
        await json.writeFile('apis.json', apis); 
        return 'Data written to apis.json' ;

    } catch(err) {
       throw err;
    } 
}


if (require.main === module) {
    const args = require('yargs')
    .option('url', {
      type: 'string',
      default: 'https://www.ardeshir.io/file.csv',
      description: 'The URL of the CSV file to fetch'
      })
      .option('web', {
      type: 'boolean',
       description: 'Start the application as a web service'
    })
    .argv;


    const dbConnString = 'postgres://postgres:postgres@localhost:5432/data';

    processCsvData(args.url, dbConnString);

    printTableApi(dbConnString);

    if (args.web) {
        const express = require('express');
        const app = express();

        app.get('/api', async (req, res) => {    
           const data = await apiHandler(); // Await the results 
           res.send(data);    
        });    

        app.listen(3000, () => {    
        console.log('Server started on port 3000');    
        });    
    }
} 

File: python3/csv_parser.py
import csv  
import json  
import psycopg2  
import requests  
from typing import List  
  
def print_table_api(db_conn_string: str) -> None:  
    conn = psycopg2.connect(db_conn_string)  
    cursor = conn.cursor()  
  
    cursor.execute("SELECT * FROM api;")  
    rows = cursor.fetchall()  
    columns = [column[0] for column in cursor.description]  
  
    print(f"Table: {'api'}")  
    for row in rows:  
        for i, col in enumerate(row):  
            print(f"{columns[i]}: {col}", end=" ")  
        print("")  
  
    cursor.close()  
    conn.close()  
  
def process_csv_data(url: str, db_conn_string: str) -> None:  
    response = requests.get(url)  
    response.raise_for_status()  
    data = list(csv.reader(response.content.decode().splitlines()[1:]))  
  
    conn = psycopg2.connect(db_conn_string)  
    cursor = conn.cursor()  
  
    cursor.execute("""  
        CREATE TABLE IF NOT EXISTS api (  
            id SERIAL PRIMARY KEY,  
            url TEXT,  
            name TEXT,  
            created INTEGER  
        );  
    """)  
  
    for row in data[1:]:  
        cursor.execute("""  
            INSERT INTO api (url, name, created)  
            VALUES (%s, %s, %s);  
        """, (row[1], row[2], row[3]))  
  
    conn.commit()  
    cursor.close()  
    conn.close()  
  
    print("Data inserted successfully!")  
  
def api_handler() -> str:  
    db_conn_string = "dbname=data user=postgres password=postgres host=localhost port=5432 sslmode=disable"  
  
    conn = psycopg2.connect(db_conn_string)  
    cursor = conn.cursor()  
  
    cursor.execute("SELECT * FROM api;")  
    rows = cursor.fetchall()  
  
    cursor.close()  
    conn.close()  
  
    columns = ["id", "url", "name", "created"]  
    apis = [dict(zip(columns, row)) for row in rows]  
  
    return json.dumps(apis)  
  
if __name__ == "__main__":  
    import argparse  
  
    parser = argparse.ArgumentParser()  
    parser.add_argument("--url", default="https://www.ardeshir.io/file.csv", help="The URL of the CSV file to fetch")  
    parser.add_argument("--web", action="store_true", help="Start the application as a web service")  
    args = parser.parse_args()  
  
    db_conn_string = "dbname=data user=postgres password=postgres host=localhost port=5432 sslmode=disable"  
  
    process_csv_data(args.url, db_conn_string)  
    print_table_api(db_conn_string)  
  
    if args.web:  
        from flask import Flask  
  
        app = Flask(__name__)  
  
        @app.route("/api")  
        def api():  
            return api_handler()  
  
        app.run()  
